hedera-node/docs/design/services/token-service/airdrops/token-cancel-airdrop.md

```diff
@@ -0,0 +1,140 @@
+     * This field is REQUIRED.
+     */
+    AccountID sender_id = 1;
+
+    /**
+     * A receiving account.<br/>
+     * This is the ID of the account that SHALL receive the airdrop.<br/>
+     * This field is REQUIRED.
+     */
+    AccountID receiver_id = 2;
+
+    oneof token_reference {
+        /**
+         * A token ID.<br/>
+         * This is the type of token for a fungible/common token airdrop.<br/>
+         * This field is REQUIRED for a fungible/common token and MUST NOT be used for a
+         * non-fungible/unique token.
+         */
+        TokenID fungible_token_type = 3;
+
+        /**
+         * The id of a single NFT, consisting of a Token ID and serial number.<br/>
+         * This is the type of token for a non-fungible/unique token airdrop.<br/>
+         * This field is REQUIRED for a non-fungible/unique token and MUST NOT be used for a
+         * fungible/common token.
+         */
+        NftID non_fungible_token = 4;
+    }
+}
+```
+
+Add new RPC to `TokenService` :
+
+```protobuf
+service TokenService {
+
+//    ...
+
+    /**
+     * Cancel one or more pending airdrops.<br/>
+     * This transaction MUST be signed by _each_ account *sending* an airdrop to be canceled.
+     */
+    rpc cancelAirdrop (Transaction) returns (TransactionResponse);
+}
+```
+
+### Fees
+
+A sender may cancel pending transfers for a low nominal fee using the new `TokenCancelAirdrop` transaction. The fee should be proportional to the number of airdrops being canceled in the transaction.
+
+An update into the `feeSchedule` file would be needed to specify that.
+
+### Services updates
+
+- Update `TokenServiceDefinition` class to include the new RPC method definition for cancelling airdrops
+- Implement new `TokenCancelAirdropHandler` class which should be invoked when the gRPC server handles `TokenCancelAirdrop` transactions. The class should be responsible for:
+    - Pure checks: validation logic based only on the transaction body itself in order to verify if the transaction is valid one
+        - Verify that the pending airdrops list contains between 1 and 10 entries, inclusive
+        - Verify that the pending airdrops list does not have any duplicate entries
+    - Pre-handle:
+        - The transaction must be signed by the account referenced by a `sender_id` for each entry in the pending airdrops list
+        - Confirm that for the given pending airdrops ids in the transaction there are corresponding pending transfers existing in state
+    - Handle:
+        - Any additional validation depending on config or state i.e. semantics checks
+        - The business logic for cancelling pending airdrops
+            - Should boil down to clearing up the pending airdrops entries from the pending airdrops state
+    - Fees calculation
+- Additional considerations:
+    - All pending transfers sent by an account must be canceled before the account can be deleted
```

Comment:
Is this tested as part of the pending state implementation or the implementation of `TokenCancel`?

Reply:
I think we should have a test for an account with pending airdrops cannot be deleted as part of the the pending state or `TokenAirdrop`.
For `TokenCancel` we can add acceptance tests for those additional considerations, namely:
- Given account with pending airdrops in state that cannot be deleted when successfull `TokenCancelAirdrop` is performed for all of the account's pending airdrops then the account can be successfully deleted.
- Given account that expires and cannot be renewed when that account is used as sender in pending airdrops then all those pending airdrops should be canceled.